src/libmongoc/tests/test-mongoc-write-concern.c

```diff
@@ -595,10 +596,135 @@ test_write_concern_wtimeout_preserved (void)
    ASSERT (bson_iter_next (&child));
    ASSERT_CMPSTR (bson_iter_key (&child), "wtimeout");
    ASSERT_CMPINT64 (bson_iter_int64 (&child), ==, INT64_MAX);
-   
+
    mongoc_write_concern_destroy (write_concern);
 }
 
+/* callback that records write concern for commands */
+static void
+write_concern_count (const mongoc_apm_command_started_t *event)
+{
+   const bson_t *command = mongoc_apm_command_started_get_command (event);
+   bson_iter_t iter, iter_rec;
+
+   if (bson_iter_init (&iter, command)) {
+      int *sent_collection_w =
+         (int *) mongoc_apm_command_started_get_context (event);
+      if (bson_iter_find_descendant (&iter, "writeConcern.w", &iter_rec)) {
+         *sent_collection_w = bson_iter_int32 (&iter_rec);
+      } else {
+         *sent_collection_w = MONGOC_WRITE_CONCERN_W_DEFAULT; /* no write
+                                                                 concern sent.
+                                                                 default used */
+      }
+   }
+}
+
+/* Addresses concerns brought up in CDRIVER-3595. This function comprises
+ * two tests. The first tests that with no txn in progress, wc is inherited
+ * from collection.  It also tests that an attempt to send an lsid with an
+ * unacknowledged wc fails.  The second tests that with a txn in progress, no
+ * wc is inherited from collection, unacknowledged or acknowledged. */
+void
+test_write_concern_inheritance_fam_txn (void *context)
+{
+   mongoc_find_and_modify_opts_t *opts = NULL;
+   bson_t *update;
+   bson_t *session_id;
+   bson_error_t error;
+   bson_t query = BSON_INITIALIZER;
+   mongoc_write_concern_t *wc;
+   mongoc_client_t *client;
+   mongoc_collection_t *collection;
+   mongoc_client_session_t *session = NULL;
+   mongoc_apm_callbacks_t *callbacks;
+   int sent_w = MONGOC_WRITE_CONCERN_W_DEFAULT;
+   bool success;
+   bool in_txn = *(bool *) context;
+
+   client = test_framework_client_new ();
+   collection = mongoc_client_get_collection (client, "db", "collection");
+
+   callbacks = mongoc_apm_callbacks_new ();
+   mongoc_apm_set_command_started_cb (callbacks, write_concern_count);
+   mongoc_client_set_apm_callbacks (client, callbacks, (void *) &sent_w);
+
+   wc = mongoc_write_concern_new ();
+   mongoc_write_concern_set_w (wc, 2);
```

Comment:
Sorry for not thinking about this before, but using `w` of `2` seems problematic for sharded cluster tests on Evergreen. 

E.g. https://evergreen.mongodb.com/task/mongo_c_driver_asan_ubuntu_test_asan_3.6_sharded_auth_nosasl_openssl_patch_368439f8410c39234f9d82d8466202176bf9d57a_5f0e8a82306615237fbcc50f_20_07_15_04_48_04 logs:

```
[2020/07/15 05:02:52.290] FAIL:/data/mci/fffe1073a527e393f08fee6bf5483918/mongoc/src/libmongoc/tests/test-mongoc-write-concern.c:681  test_write_concern_inheritance_fam_txn()
[2020/07/15 05:02:52.290]   success
[2020/07/15 05:02:52.290]   Write Concern error: Not enough data-bearing nodes at sh01
```

As well as standalones (which support sessions).

Since the default behavior is to not send any write concern, I think this can be set to 1 without adding any ambiguity.